modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteMergeJoin.java

```diff
@@ -0,0 +1,634 @@
+                    else {
+                        double leftDistinct = Util.first(
+                            mq.getDistinctRowCount(getLeft(), ImmutableBitSet.of(leftKeys), null), left1);
+                        double rightDistinct = Util.first(
+                            mq.getDistinctRowCount(getRight(), ImmutableBitSet.of(rightKeys), null), right1);
+                        double leftCardinality = leftDistinct / left1;
+                        double rightCardinality = rightDistinct / right1;
+                        double rowsCount = (Math.min(left1, right1) / (leftCardinality * rightCardinality)) * selectivity;
+                        JoinRelType type = getJoinType();
+                        if (type == LEFT)
+                            rowsCount += left1;
+                        else if (type == RIGHT)
+                            rowsCount += right1;
+                        else if (type == JoinRelType.FULL)
+                            rowsCount += left1 + right1;
+                        result = rowsCount;
+                    }
+                }
+            }
+        }
+
+        return Util.first(result, 1D);
+    }
+
+    /** {@inheritDoc} */
+    @Override public RelOptCost computeSelfCost(RelOptPlanner planner, RelMetadataQuery mq) {
+        double rowCount = mq.getRowCount(this);
+
+        // Joins can be flipped, and for many algorithms, both versions are viable
+        // and have the same cost. To make the results stable between versions of
+        // the planner, make one of the versions slightly more expensive.
+        switch (joinType) {
+            case SEMI:
+            case ANTI:
+                // SEMI and ANTI join cannot be flipped
+                break;
+            case RIGHT:
+                rowCount = RelMdUtil.addEpsilon(rowCount);
+                break;
+            default:
+                if (RelNodes.COMPARATOR.compare(left, right) > 0)
+                    rowCount = RelMdUtil.addEpsilon(rowCount);
+        }
+
+        final double rightRowCount = right.estimateRowCount(mq);
+        final double leftRowCount = left.estimateRowCount(mq);
+
+        if (Double.isInfinite(leftRowCount))
+            rowCount = leftRowCount;
+        if (Double.isInfinite(rightRowCount))
+            rowCount = rightRowCount;
+
+        RelDistribution.Type type = distribution().getType();
+
+        if (type == BROADCAST_DISTRIBUTED || type == SINGLETON)
+            rowCount = RelMdUtil.addEpsilon(rowCount);
+
+        return planner.getCostFactory().makeCost(rowCount, 0, 0);
+    }
+
+    private static <T> List<T> maxPrefix(List<T> seq, Collection<T> elems) {
+        List<T> res = new ArrayList<>();
+
+        Set<T> elems0 = new HashSet<>(elems);
+
+        for (T e : seq) {
+            if (!elems0.remove(e))
+                break;
+
+            res.add(e);
+        }
+
+        return res;
+    }
+
+    private static <T> boolean isPrefix(List<T> seq, Collection<T> elems) {
```

Comment:
1. no javadoc
2. strange contract - collection as input and set as inner struct
3. if you need real prefix comparison of two collections you don`t need additional Set, i hope.

Reply:
I add the javadoc. Hope it helps to get the idea of the method